### ISSUE TITLE
Redis tracing

### ISSUE THREAD

**Issue by lewispb** [getsentry/sentry-ruby#1693]
**Describe the idea**

Similar to the existing Sentry Performance support for Active Record database tracing, we've found it useful to view the time taken to execute Redis commands.

I've created a simple custom Redis instrumenter and tracing subscriber. With this enabled, we can inspect traces including Redis commands:

<img width="1432" alt="sentry-performance-with-redis" src="https://user-images.githubusercontent.com/1773614/150431894-e0277ab3-0f92-488a-a94b-0e8422eb3406.png">

**Why do you think it's beneficial to most of the users**

Redis is a popular and widely-used database.

**Possible implementation**

We would add the following to the "sentry-rails" gem:

- A custom instrumenter, inspired by the [Redis Rails Instrumentation gem](https://github.com/lautis/redis-rails-instrumentation)
- A subscriber to the newly generated Active Support Notifications

I considered a new gem "sentry-redis", but it appears that the Performance product with tracing is only available for Rails apps. 

### Instrumentation example

```rb
class Redis
  module Instrumentation
    def logging(commands, &block)
      ActiveSupport::Notifications.instrument("command.redis", commands: commands, server: "#{host}:#{port}/#{db}") do
        return super(commands, &block)
      end
    end
  end
end

Redis::Client.prepend(Redis::Instrumentation)
```

### Subscriber example

The first command and key is instrumented, but not the value.

```rb
require "sentry/rails/tracing/abstract_subscriber"

module Sentry
  module Rails
    module Tracing
      class RedisSubscriber < AbstractSubscriber
        def self.subscribe!
          subscribe_to_event([ "command.redis" ]) do |event_name, duration, payload|
            description = payload.dig(:commands, 0).take(2).join(" ")

            record_on_current_span(
              op: event_name,
              start_timestamp: payload[START_TIMESTAMP_NAME],
              description: description,
              duration: duration
            ) do |span|
              span.set_data(:redis_server, payload[:server])
            end
          end
        end
      end
    end
  end
end
```

I'm happy to implement this if the approach is acceptable. Thanks!

**Comment by st0012** [comment]
Thanks for the proposal, I think it'd be a great feature to add!

> I considered a new gem "sentry-redis", but it appears that the Performance product with tracing is only available for Rails apps.

That's not exactly correct. You can do performance tracing even without `ActiveSupport::Notifications` or Rails. It surely is not as convenient as subscribing an event though. You can use the SDK's [net-http tracing](https://github.com/getsentry/sentry-ruby/blob/9380fbc521445796f8d4a4d86ad2ef3bdbaa5719/sentry-ruby/lib/sentry/net/http.rb#L30-L36) as an example.

Ideally, `sentry-redis` would be a great separation. But since `redis` is a very common component for our users, I think we can have this feature directly in `sentry-ruby`. We just need to handle the dependency expectation more carefully.

I can help you on the implementation or implement myself, so let me know how you want to proceed on this 🙂

**Comment by lewispb** [comment]
That's cool thanks @st0012. Glad you like the proposal. I didn't know about the tracing with AS::N but that sounds like a better path to me. 👍

I made a start on implementation (with a `redis_logger` breadcrumbs logger) according to your guidance here: https://github.com/getsentry/sentry-ruby/pull/1697

I'll work on it a bit more this week but of course if you'd like to leave any early feedback that'd be welcome. There's a few things left for me to figure out. I've written a few questions on the PR. Thanks again!

**Comment by st0012** [comment]
Closed via #1697
